### PR TITLE
Finetuning/transfer learning

### PR DESCRIPTION
**Pull request by wyli**
## Status
**WORK IN PROGRESS**

## Description
implementing interfaces for transfer learning,
see https://github.com/NifTK/NiftyNet/issues/37 and
https://stackoverflow.com/questions/50768864/how-do-i-implement-transfer-learning-in-niftynet
This PR looks at finetuning from a checkpoint saved by niftynet (therefore name scopes in the file are consistent with niftynet's network definition), checkpoints from other tensorflow packages are not considered here.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Todos
- [ ] command for printing all trainable variable names (with optional regular expression matching)
- [x] randomly initialise a subset of variables, this subset is created by regex name matching
- [x] restore (from an existing checkpoint) and continue updating a subset of the variables
    - [ ] if the optimisation method changed, deal with method-specific variables (e.g. momentum)?
- [x] restore (from an existing checkpoint) and freeze trained weights for the rest of the variables
- [x] save all trainable variables after training
- [ ] configuration parameters for finetuning, variable name regex, 
- [ ] unit tests
- [ ] a demo/tutorial
- [ ] checkpoint's scope names might depend on the version of NN, check the compatibility? (or preprocess the checkpoints)
- [ ] deal with batch norm and dropout layers (editing networks to remove batch norm variables?)

## Impacted Areas in Application
List general components of the application that this PR will affect:
* segmentation application
* regression application
* classification application
* niftynet/engine
* new configuration parameters

### CODE REVIEW

**Comment by wyli** [comment]
an improved version is merged via PR https://github.com/NifTK/NiftyNet/pull/290

**Comment by koriavinash1** [comment]
@wyli Are demo/tutorials out?

**Comment by wyli** [comment]
@koriavinash1 yes https://niftynet.readthedocs.io/en/dev/transfer_learning.html
